internal/workflow/reconciler.go

```diff
@@ -2,171 +2,72 @@ package workflow
-
-	data := make(map[string]interface{})
-	for key, val := range stored.Spec.HardwareMap {
-		data[key] = val
-	}
-
-	var hardware v1alpha1.Hardware
-	err := r.client.Get(ctx, ctrlclient.ObjectKey{Name: stored.Spec.HardwareRef, Namespace: stored.Namespace}, &hardware)
-	if err != nil && !errors.IsNotFound(err) {
-		logger.Error(err, "error getting Hardware object in processNewWorkflow function")
-		return reconcile.Result{}, err
-	}
-
-	if stored.Spec.HardwareRef != "" && errors.IsNotFound(err) {
-		logger.Error(err, "hardware not found in processNewWorkflow function")
-		return reconcile.Result{}, fmt.Errorf(
-			"hardware not found: name=%v; namespace=%v",
-			stored.Spec.HardwareRef,
-			stored.Namespace,
-		)
-	}
-
-	if err == nil {
-		contract := toTemplateHardwareData(hardware)
-		data["Hardware"] = contract
-	}
-
-	tinkWf, err := renderTemplateHardware(stored.Name, ptr.StringValue(tpl.Spec.Data), data)
-	if err != nil {
-		return reconcile.Result{}, err
-	}
-
-	// populate Task and Action data
-	stored.Status = *YAMLToStatus(tinkWf)
 
-	stored.Status.State = v1alpha1.WorkflowStatePending
-	return reconcile.Result{}, nil
-}
-
-// templateHardwareData defines the data exposed for a Hardware instance to a Template.
-type templateHardwareData struct {
-	Disks []string
-}
-
-// toTemplateHardwareData converts a Hardware instance of templateHardwareData for use in template
-// rendering.
-func toTemplateHardwareData(hardware v1alpha1.Hardware) templateHardwareData {
-	var contract templateHardwareData
-	for _, disk := range hardware.Spec.Disks {
-		contract.Disks = append(contract.Disks, disk.Device)
+	// TODO(chrisdoherty)
+	if !wrkflw.DeletionTimestamp.IsZero() {
+		return reconcile.Result{}, nil
 	}
-	return contract
-}
 
-func (r *Reconciler) processRunningWorkflow(_ context.Context, stored *v1alpha1.Workflow) reconcile.Result {
-	// Check for global timeout expiration
-	if r.nowFunc().After(stored.GetStartTime().Add(time.Duration(stored.Status.GlobalTimeout) * time.Second)) {
-		stored.Status.State = v1alpha1.WorkflowStateTimeout
+	rc := internal.ReconciliationContext{
+		Client:   r.client,
+		Log:      logger,
+		Workflow: wrkflw.DeepCopy(),
 	}
 
-	// check for any running actions that may have timed out
-	for ti, task := range stored.Status.Tasks {
-		for ai, action := range task.Actions {
-			// A running workflow task action has timed out
-			if action.Status == v1alpha1.WorkflowStateRunning && action.StartedAt != nil &&
-				r.nowFunc().After(action.StartedAt.Add(time.Duration(action.Timeout)*time.Second)) {
-				// Set fields on the timed out action
-				stored.Status.Tasks[ti].Actions[ai].Status = v1alpha1.WorkflowStateTimeout
-				stored.Status.Tasks[ti].Actions[ai].Message = "Action timed out"
-				stored.Status.Tasks[ti].Actions[ai].Seconds = int64(r.nowFunc().Sub(action.StartedAt.Time).Seconds())
-				// Mark the workflow as timed out
-				stored.Status.State = v1alpha1.WorkflowStateTimeout
-			}
+	// Always attempt to patch.
```

Comment:
Why always attempt a patch?

Reply:
The rationale is that you do as much as possible to bring the resource to the desired state and you shouldn't update the `.Status` state if nothing changed so always patching ensures we always try to persist the updates.